cache/cache.go

```diff
@@ -0,0 +1,443 @@
+	// default TTL as the new TTL, calculated from the current time.
+	// If the key does not exist, the operation should have no effect.
+	Renew(key string)
+
+	// GetItems retrieves all unexpired items in the cache.
+	// It returns a map of string keys to Item values
+	// representing the current state of the cache.
+	GetItems() map[string]Item
+
+	// GetAllItems retrieves all items in the cache regardless of expiration.
+	// It returns a map of string keys to Item values
+	// representing the current state of the cache.
+	GetAllItems() map[string]Item
+
+	// Get retrieves the value associated with the given key.
+	// It returns the value and true if the key exists and has not expired according
+	// to the implementation's clock and expiration logic.
+	// Otherwise, it returns nil and false.
+	Get(key string) (any, bool)
+
+	// GetTTL retrieves the remaining time-to-live (TTL) for the given key.
+	// It returns the remaining duration and true if the key exists and has not expired.
+	// Otherwise, it returns 0 and false.
+	GetTTL(key string) (time.Duration, bool)
+
+	// Delete removes a key-value pair from the cache.
+	// If the key does not exist, the operation should have no effect.
+	Delete(key string)
+
+	// Clear removes all key-value pairs from the cache.
+	Clear()
+
+	// DeleteExpired manually triggers the removal of all expired items from the cache
+	// based on the implementation's clock and expiration logic.
+	DeleteExpired()
+
+	// Stop stops the garbage collector if it is running.
+	// This is typically called when the cache is no longer needed.
+	// It is important to call this function to avoid resource leaks
+	// and ensure that the goroutine is properly cleaned up.
+	Stop()
+
+	// GetGarbageCollector returns the current garbage collector.
+	GetGarbageCollector() *garbageCollector
+
+	// SetGarbageCollector sets the garbage collector for the cache.
+	SetGarbageCollector(gc *garbageCollector)
+
+	// GetClock returns the clock used by the cache for time-related operations.
+	GetClock() Clock
+
+	// SetClock sets the clock for the cache.
+	SetClock(clock Clock)
+}
+
+// cache represents an in-memory key-value store with time-based expiration.
+// It is safe for concurrent use.
+type cache struct {
+	// items holds the cache data.
+	items map[string]Item
+	// mu provides concurrency control for safe access to items.
+	mu sync.RWMutex
+	// garbageCollector manages the periodic removal of expired items.
+	garbageCollector *garbageCollector
+	// cacheInterval specifies the default duration for items added without a specific TTL.
+	cacheInterval time.Duration
+	// garbageCollectorInterval defines how often the garbage collector runs.
+	garbageCollectorInterval time.Duration
+	// allowOverwrites determines if Add operations can overwrite existing, non-expired keys.
+	allowOverwrites bool
+	// clock provides an abstraction for time, useful for testing.
+	clock Clock
```

Comment:
I would give serious consideration to these features. Do we really need to have that level of control? `cacheInterval`/`ttl` is probably fine, but I just don't know about the others. It doesn't go well with the 'simple' label.